src/MonoTorrent/MonoTorrent.BEncoding/BEncodeDecoder.cs

```diff
@@ -0,0 +1,164 @@
+        {
+            BEncodedString key = null;
+            BEncodedValue value = null;
+            BEncodedString oldkey = null;
+
+            int read;
+            var dictionary = new BEncodedDictionary ();
+            while ((read = reader.ReadByte ()) != -1 && read != 'e') {
+                key = DecodeString (reader, read - '0');         // keys have to be BEncoded strings
+
+                if (oldkey != null && oldkey.CompareTo (key) > 0)
+                    if (strictDecoding)
+                        throw new BEncodingException (
+                            $"Illegal BEncodedDictionary. The attributes are not ordered correctly. Old key: {oldkey}, New key: {key}");
+
+                oldkey = key;
+                value = Decode (reader);                     // the value is a BEncoded value
+                dictionary.Add (key, value);
+            }
+
+            if (read != 'e')                                    // remove the trailing 'e'
+                throw new BEncodingException ("Invalid data found. Aborting");
+            return dictionary;
+        }
+
+        static BEncodedList DecodeList (RawReader reader)
+        {
+            var list = new BEncodedList ();
+            int read;
+            while ((read = reader.ReadByte ()) != -1 && read != 'e')
+                list.Add (Decode (reader, read));
+
+            if (read != 'e')                            // Remove the trailing 'e'
+                throw new BEncodingException ("Invalid data found. Aborting");
+
+            return list;
+        }
+
+        static BEncodedNumber DecodeNumber (RawReader reader)
+        {
+            int sign = 1;
+            long result = 0;
+            int val = reader.ReadByte ();
+            if (val == '-') {
+                sign = -1;
+                val = reader.ReadByte ();
+            }
+
+            do {
+                if (val < '0' || val > '9')
+                    throw new BEncodingException ("Invalid number found.");
+                result = result * 10 + (val - '0');
+            } while ((val = reader.ReadByte ()) != 'e' && val != -1);
+
+            if (val == -1)        //remove the trailing 'e'
+                throw new BEncodingException ("Invalid data found. Aborting.");
+
+            result *= sign;
+            return result;
+        }
+
+        static BEncodedString DecodeString (RawReader reader, int length)
+        {
+            int read;
+            while ((read = reader.ReadByte ()) != -1 && read != ':') {
+                if (read < '0' || read > '9')
+                    throw new BEncodingException ($"Invalid BEncodedString. Length was '{length}' instead of a number");
+                length = length * 10 + (read - '0');
+            }
+
+            if (read != ':')
+                throw new BEncodingException ("Invalid data found. Aborting");
+
+            var bytes = new byte[length];
```

Comment:
Are these strings really long? Otherwise maybe ArrayPool<byte> could be used here.

Reply:
The byte[] is retained in the BEncodedString object forever. A BEncodedString is a raw byte[] which can sometimes be converted into a valid UTF-8 string. If a UTF-8 string is needed it is generated on the fly and not retained